Prototype Priavate AI integration for NeMo-Guardrails (#2)
privateai/NeMo-Guardrails: docs/user_guides/community/privateai.md

# Private AI Integration

[Private AI](https://docs.private-ai.com) allows you to detect and mask Personally Identifiable Information (PII) in your data. This integration enables NeMo Guardrails to use Private AI for PII detection in input, output and retrieval flows.

## Setup

> 1. Ensure you have access to a Private AI API endpoint running locally or in the cloud. To get started with the cloud version, you can use the [Private AI Portal](https://portal.private-ai.com). For containerized deployments check this [Quickstart Guide](https://docs.private-ai.com/quickstart/).

Reviewer: ```suggestion
1. Ensure you have access to a Private AI API endpoint running locally or in the cloud. To get started with the cloud version, you can use the [Private AI Portal](https://portal.private-ai.com). For containerized deployments, check out this [Quickstart Guide](https://docs.private-ai.com/quickstart/).
```

> 2. Configure your `config.yml` file to include the Private AI settings:

Reviewer: ```suggestion
2. Update your `config.yml` file to include the Private AI settings:
```
Maybe this sounds better? Configure the config sounded interesting to me, but this is a suggestion, not a dealbreaker.

```yaml
rails:
  config:
    privateai:

>       server_endpoint: http://your-privateai-api-endpoint/process/text  # Replace with your Private AI API endpoint

Reviewer: ```suggestion
      server_endpoint: http://your-privateai-api-endpoint/process/text  # Replace this with your Private AI process text endpoint
```

      api_key: your-privateai-api-key  # Optional: Not needed for the containerized version
      input:
        entities:
          - NAME_FAMILY
          - LOCATION_ADDRESS_STREET
          - EMAIL_ADDRESS
      output:
        entities:
          - NAME_FAMILY
          - LOCATION_ADDRESS_STREET

>           - EMAIL_ADDRESS

Reviewer: Do we enable all entities by default? Maybe we can talk about that here? Otherwise, this example gives the feeling that you need to list all of our entities if you want to enable all of them.

Author: That's a nice catch. I enabled all entities by default and added a comment here.

  input:
    flows:
      - detect pii on input
  output:
    flows:

>       - detect pii on output

Reviewer: Do we define the name of the flows?

Author: Yes, here: https://github.com/privateai/NeMo-Guardrails/blob/privateai-prototype/nemoguardrails/library/privateai/flows.v1.co

```

Replace `http://your-privateai-api-endpoint/process/text` with your actual Private AI API endpoint.

> 3. You can customize the `entities` list under both `input` and `output` to include the PII types you want to detect. A full list of supported entities can be found [here](https://docs.private-ai.com/entities).

Reviewer: Please refer to the previous comments about these.

## Usage

Once configured, the Private AI integration will automatically:

1. Detect PII in user inputs before they are processed by the LLM.
2. Detect PII in LLM outputs before they are sent back to the user.
3. Detect PII in retrieved chunks before they are sent to the LLM.

The `detect_pii` action in `nemoguardrails/library/privateai/actions.py` handles the PII detection process.

## Customization

You can customize the PII detection behavior by modifying the `entities` lists in the `config.yml` file. Refer to the Private AI documentation for a complete list of [supported entity types](https://docs.private-ai.com/entities).

## Error Handling

If the Private AI detection API request fails, the system will assume PII is present as a precautionary measure.

## Notes

> - Ensure that your Private AI API endpoint is properly set up and accessible from your NeMo Guardrails environment.

Reviewer: ```suggestion
- Ensure that your Private AI process text endpoint is properly set up and accessible from your NeMo Guardrails environment.
```
